File: VB/WebDAVServer.SqlStorage.HttpListener/WebSocketsService.vb
Imports System
Imports System.Collections.Concurrent
Imports System.Net.WebSockets
Imports System.Text
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Web.Script.Serialization

''' <summary>
''' Notifies client about changes in WebDAV items.
''' </summary>
Public Class WebSocketsService

    ''' <summary>
    ''' Singleton instance of <see cref="WebSocketsService"/> .
    ''' </summary>
    Public Shared ReadOnly Service As WebSocketsService = New WebSocketsService()

    ''' <summary>
    ''' Creates instance of this class.
    ''' </summary>
    Private Sub New()
    End Sub

    ''' <summary>
    ''' Dictionary which contains connected clients.
    ''' </summary>
    Private ReadOnly clients As ConcurrentDictionary(Of Guid, WebSocket) = New ConcurrentDictionary(Of Guid, WebSocket)()

    ''' <summary>
    ''' Adds client to connected clients dictionary.
    ''' </summary>
    ''' <param name="client">Current client.</param>
    ''' <returns>Client guid in the dictionary.</returns>
    Public Function AddClient(client As WebSocket) As Guid
        Dim clientId As Guid = Guid.NewGuid()
        clients.TryAdd(clientId, client)
        Return clientId
    End Function

    ''' <summary>
    ''' Removes client from connected clients dictionary.
    ''' </summary>
    ''' <param name="clientId">Client guid in the dictionary.</param>
    Public Sub RemoveClient(clientId As Guid)
        Dim client As WebSocket
        clients.TryRemove(clientId, client)
    End Sub

    ''' <summary>
    ''' Notifies client that file/folder was created.
    ''' </summary>
    ''' <param name="itemPath">file/folder.</param>
    ''' <returns></returns>
    Public Async Function NotifyCreatedAsync(itemPath As String) As Task
        Await SendMessage(itemPath, "created")
    End Function

    ''' <summary>
    ''' Notifies client that file/folder was updated.
    ''' </summary>
    ''' <param name="itemPath">file/folder path.</param>
    ''' <returns></returns>
    Public Async Function NotifyUpdatedAsync(itemPath As String) As Task
        Await SendMessage(itemPath, "updated")
    End Function

    ''' <summary>
    ''' Notifies client that file/folder was deleted.
    ''' </summary>
    ''' <param name="itemPath">file/folder path.</param>
    ''' <returns></returns>
    Public Async Function NotifyDeletedAsync(itemPath As String) As Task
        Await SendMessage(itemPath, "deleted")
    End Function

    ''' <summary>
    ''' Notifies client that file/folder was locked.
    ''' </summary>
    ''' <param name="itemPath">file/folder path.</param>
    ''' <returns></returns>
    Public Async Function NotifyLockedAsync(itemPath As String) As Task
        Await SendMessage(itemPath, "locked")
    End Function

    ''' <summary>
    ''' Notifies client that file/folder was unlocked.
    ''' </summary>
    ''' <param name="itemPath">file/folder path.</param>
    ''' <returns></returns>
    Public Async Function NotifyUnLockedAsync(itemPath As String) As Task
        Await SendMessage(itemPath, "unlocked")
    End Function

    ''' <summary>
    ''' Notifies client that file/folder was moved.
    ''' </summary>
    ''' <param name="itemPath">old file/folder path.</param>
    ''' <param name="targetPath">new file/folder path.</param>
    ''' <returns></returns>
    Public Async Function NotifyMovedAsync(itemPath As String, targetPath As String) As Task
        itemPath = itemPath.Trim("/"c)
        targetPath = targetPath.Trim("/"c)
        Dim notifyObject As MovedNotification = New MovedNotification With {.ItemPath = itemPath,
                                                                      .TargetPath = targetPath,
                                                                      .EventType = "moved"}
        For Each client As WebSocket In clients.Values
            If client.State = WebSocketState.Open Then
                Await client.SendAsync(New ArraySegment(Of Byte)(Encoding.UTF8.GetBytes(New JavaScriptSerializer().Serialize(notifyObject))), WebSocketMessageType.Text, True, CancellationToken.None)
            End If
        Next
    End Function

    ''' <summary>
    ''' Sends message about file/folder operation.
    ''' </summary>
    ''' <param name="itemPath">File/Folder path.</param>
    ''' <param name="operation">Operation name: created/updated/deleted/moved</param>
    ''' <returns></returns>
    Public Async Function SendMessage(itemPath As String, operation As String) As Task
        itemPath = itemPath.Trim("/"c)
        Dim notifyObject As Notification = New Notification With {.ItemPath = itemPath,
                                                            .EventType = operation
                                                            }
        For Each client As WebSocket In clients.Values
            If client.State = WebSocketState.Open Then
                Await client.SendAsync(New ArraySegment(Of Byte)(Encoding.UTF8.GetBytes(New JavaScriptSerializer().Serialize(notifyObject))), WebSocketMessageType.Text, True, CancellationToken.None)
            End If
        Next
    End Function
End Class

''' <summary>
''' Holds notification information.
''' </summary>
Public Class Notification

    ''' <summary>
    ''' Represents file/folder path.
    ''' </summary>
    Public Property ItemPath As String = String.Empty

    ''' <summary>
    ''' Represents event type.
    ''' </summary>
    Public Property EventType As String = String.Empty
End Class

''' <summary>
''' Holds notification moved information.
''' </summary>
Public Class MovedNotification
    Inherits Notification

    ''' <summary>
    ''' Represents target file/folder path.
    ''' </summary>
    Public Property TargetPath As String = String.Empty
End Class

File: CS/AzureDataLakeStorage/WebSocketsService.cs
using System;
using System.Collections.Concurrent;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

using Newtonsoft.Json;


namespace AzureDataLakeStorage
{
    /// <summary>
    /// Notifies client about changes in WebDAV items.
    /// </summary>
    public class WebSocketsService
    {

        /// <summary>
        /// Dictionary which contains connected clients.
        /// </summary>
        private readonly ConcurrentDictionary<Guid, WebSocket> clients = new ConcurrentDictionary<Guid, WebSocket>();

        /// <summary>
        /// Adds client to connected clients dictionary.
        /// </summary>
        /// <param name="client">Current client.</param>
        /// <returns>Client guid in the dictionary.</returns>
        public Guid AddClient(WebSocket client)
        {
            Guid clientId = Guid.NewGuid();
            clients.TryAdd(clientId, client);
            return clientId;
        }

        /// <summary>
        /// Removes client from connected clients dictionary.
        /// </summary>
        /// <param name="clientId">Client guid in the dictionary.</param>
        public void RemoveClient(Guid clientId)
        {
            clients.TryRemove(clientId, out _);
        }

        /// <summary>
        /// Notifies client that content in the specified folder has been changed. 
        /// Called when one of the following events occurs in the specified folder: file or folder created, file or folder updated, file deleted.
        /// </summary>
        /// <param name="folderPath">Content of this folder was modified.</param>
        /// <returns></returns>
        public async Task NotifyRefreshAsync(string folderPath)
        {
            folderPath = folderPath.Trim('/');
            Notification notifyObject = new Notification
            {
                FolderPath = folderPath,
                EventType = "refresh"
            };
            foreach (WebSocket client in clients.Values)
            {
                if (client.State == WebSocketState.Open)
                {
                    await client.SendAsync(new ArraySegment<byte>(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(notifyObject))), WebSocketMessageType.Text, true, CancellationToken.None);
                }
            }
        }

        /// <summary>
        /// Notifies client that folder was deleted.
        /// </summary>
        /// <param name="folderPath">Folder that was deleted.</param>
        /// <returns></returns>
        public async Task NotifyDeleteAsync(string folderPath)
        {
            folderPath = folderPath.Trim('/');
            Notification notifyObject = new Notification
            {
                FolderPath = folderPath,
                EventType = "delete"
            };
            foreach (WebSocket client in clients.Values)
            {
                if (client.State == WebSocketState.Open)
                {
                    await client.SendAsync(new ArraySegment<byte>(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(notifyObject))), WebSocketMessageType.Text, true, CancellationToken.None);
                }
            }
        }
    }

    /// <summary>
    /// Holds notification information.
    /// </summary>
    public class Notification
    {
        /// <summary>
        /// Represents notification data.
        /// </summary>
        public string FolderPath { get; set; } = string.Empty;

        /// <summary>
        /// Represents event type.
        /// </summary>
        public string EventType { get; set; } = string.Empty;
    }
}
